Re-implement map_over_datasets using group_subtrees

shoyer opened the pull request:
Copied from https://github.com/shoyer/xarray/pull/2

Fixes #9643

To recap:

- It is implemented using zip_subtrees, which means it should properly
handle DataTrees where the nodes are defined in a different order. (closing https://github.com/pydata/xarray/issues/9643)
- For simplicity, I removed handling of `**kwargs`, in order to preserve
some flexibility for adding keyword arugments.
- I removed automatic skipping of empty nodes, because there are almost
assuredly cases where that would make sense. This could be restored
with a option keyword arugment.

To do:
- [x] change `map_over_datasets` from being called like `map_over_datasets(func)(*args)` to `map_over_datasets(func, *args)`, which would be more consistent with apply_ufunc.
- [x] create an alternative `group_subtrees` interface that yields tuples `(path, nodes)` , which will hopefully make it harder to make bugs like https://github.com/pydata/xarray/pull/9626

shoyer commented on xarray/core/datatree_mapping.py:
```diff
@@ -125,110 +123,55 @@ def map_over_datasets(func: Callable) -> Callable:
+                    node_dataset_args.insert(i, arg)
+
+            path = (
+                "/"
```
From @TomNicholas's review:

> Are you sure that's intended? I think this might effectively orphan the result (possibly similar to https://github.com/pydata/xarray/pull/9626#pullrequestreview-2369600925).

shoyer replied:
I've switching to specifying relative paths returned via `group_subtrees`

TomNicholas replied:
That seems like a much more robust pattern in general.

shoyer commented:
This has gotten a bit bigger. I've also:

1. reimplemented `DataTree.isomorphic` to use `group_subtrees`
2. added a new iterator `subtree_with_path`
3. reimplemented `filter` and `match` using `subtree_with_path` (to fix a bug when applying operations to a subtree)
4. updated the documentation

shoyer commented:
Anyways this is ready for review!

TomNicholas commented on xarray/core/treenode.py:
```diff
@@ -414,6 +417,28 @@ def subtree(self: Tree) -> Iterator[Tree]:
+            path, node = queue.popleft()
+            yield path, node
+            queue.extend(
+                (os.path.join(path, name), child)
```
I bet this is what's causing the test failures on Windows. Paths between nodes cannot be treated the same as filesystem paths! That's why I made [`NodePath`](https://github.com/pydata/xarray/blob/b9780e7a32b701736ebcf33d9cb0b380e92c91d5/xarray/core/treenode.py#L29), so you could just use that here to fix it.

TomNicholas replied:
Another reason to hide path manipulation logic behind an abstraction is so that we can generalize it later to handle / reject `Hashables`. https://github.com/pydata/xarray/issues/8836#issuecomment-2342295573

TomNicholas replied:
You got there just as I was reviewing! So this comment should be fixed by https://github.com/pydata/xarray/pull/9636/commits/1f07b63ec44a3a6703e0f67e6485c2e3e4d48df4

TomNicholas commented on doc/user-guide/hierarchical-data.rst:
```diff
@@ -362,21 +362,26 @@ This returns an iterable of nodes, which yields them in depth-first order.
+Similarly, :py:class:`~xarray.DataTree.subtree_with_keys` returns an iterable of
+relative paths and corresponding nodes.
 
+A very useful pattern is to iterate over :py:class:`~xarray.DataTree.subtree_with_keys`
+to manipulate nodes however you wish, then rebuild a new tree using
+:py:meth:`xarray.DataTree.from_dict()`.
```
So we have a new property:

```python
assert DataTree.from_dict(dt.subtree_with_keys()) == dt
```

Is there a good place we could document these properties?

TomNicholas commented on doc/user-guide/hierarchical-data.rst:
```diff
@@ -362,21 +362,26 @@ This returns an iterable of nodes, which yields them in depth-first order.
 You can see this tree is similar to the ``dt`` object above, except that it is missing the empty nodes ``a/c`` and ``a/c/d``.
 
-(If you want to keep the name of the root node, you will need to add the ``name`` kwarg to :py:class:`~xarray.DataTree.from_dict`, i.e. ``DataTree.from_dict(non_empty_nodes, name=dt.root.name)``.)
+(If you want to keep the name of the root node, you will need to add the ``name`` kwarg to :py:class:`~xarray.DataTree.from_dict`, i.e. ``DataTree.from_dict(non_empty_nodes, name=dt.name)``.)
```
Good catch!

TomNicholas commented on xarray/core/datatree.py:
```diff
@@ -1279,12 +1267,12 @@ def equals(self, other: DataTree) -> bool:
             return False
 
         return all(
             node.dataset.equals(other_node.dataset)
```
A little clearer intent?
```suggestion
            # do not check that inherited coordinates are defined on the same level
            node.to_dataset(inherit=True).equals(other_node.to_dataset(inherit=True))
```

TomNicholas commented on xarray/core/datatree.py:
```diff
@@ -1418,18 +1406,15 @@ def map_over_datasets(
         -------
         subtrees : DataTree, tuple of DataTrees
             One or more subtrees containing results from applying ``func`` to the data at each node.
         """
```
```suggestion
        
        See Also
        --------
        map_over_datasets
        """
```

TomNicholas commented on xarray/core/datatree_mapping.py:
```diff
@@ -1,236 +1,129 @@
 
-    pass
 
+@overload
```
I think this is true?

```suggestion
# special-case for binary ops. Unfortunately there is no easy way to type hint that the number of output trees will always equal the number of return values of the Callable.
@overload
```

TomNicholas commented on xarray/core/datatree_mapping.py:
```diff
@@ -1,236 +1,129 @@
-        raise TypeError(f"Argument `a` is not a tree, it is of type {type(a)}")
-    if not isinstance(b, TreeNode):
-        raise TypeError(f"Argument `b` is not a tree, it is of type {type(b)}")
+    Decorator which turns a function which acts on (and returns) Datasets into one
+    which acts on and returns DataTrees.
```
```suggestion
```

TomNicholas commented on xarray/core/datatree_mapping.py:
```diff
@@ -1,236 +1,129 @@
+    store the results.
 
-    diff = diff_treestructure(a, b, require_names_equal=require_names_equal)
+    The function will be applied to any data-containing dataset stored in any of
```
I don't think you're doing `if node.has_data` any more?
```suggestion
    The function will be applied to any dataset stored in any of
```

TomNicholas commented on xarray/core/datatree_mapping.py:
```diff
@@ -1,236 +1,129 @@
+    returned value that is one of these types will be stacked into a separate
+    tree before returning all of them.
 
+    ``map_over_datasets`` is essentially syntactic sugar for the combination of
+    ``group_subtrees`` and ``DataTree.from_dict``. For example, in the case of
```
This is really nice and clear.

TomNicholas commented on xarray/core/datatree_mapping.py:
```diff
@@ -1,236 +1,129 @@
+    # Store tuples of results in a dict because we don't yet know how many trees we need to rebuild to return
+    out_data_objects: dict[str, Dataset | None | tuple[Dataset | None, ...]] = {}
+
+    tree_args = [arg for arg in args if isinstance(arg, DataTree)]
+    name = result_name(tree_args)
```
```suggestion
    tree_args = [arg for arg in args if isinstance(arg, DataTree)]
    
    if tree_args == []:
       raise ValueError("Must pass at least one `DataTree`")
    
    name = result_name(tree_args)
```

shoyer replied:
You already get an error message like this (which is tested):
```
>>> xr.map_over_datasets(lambda x: x, "dt")
TypeError: must pass at least one tree object
```

TomNicholas commented on xarray/core/datatree_mapping.py:
```diff
@@ -1,236 +1,129 @@
+        `func(*args: Dataset) -> Union[Dataset, tuple[Dataset, ...]]`.
 
         (i.e. func must accept at least one Dataset and return at least one Dataset.)
         Function will not be applied to any nodes without datasets.
```
Again, is that still true?

TomNicholas commented on xarray/core/treenode.py:
```diff
@@ -778,42 +805,79 @@ def _path_to_ancestor(self, ancestor: NamedNode) -> NodePath:
 
     # https://en.wikipedia.org/wiki/Breadth-first_search#Pseudocode
-    queue = collections.deque([trees])
+    queue = collections.deque([("", trees)])
```
I guess this answers my question about why `bool(path)` would ever be `False`. But why are we doing this exactly?

shoyer replied:
We could equivalently return `""`, `"."`, `"/"` or `./` for the top-level node. Or we could make `None` a sentinel value recognized by `from_dict`.

Given that the others are relative paths, I would lean towards the strings `""` or `"."`. I guess `"."` might make more sense since it's the canonical path returned by `NodePath`?

TomNicholas replied:
> We could equivalently return "", ".", "/" or ./ for the top-level node

We shouldn't return `"/"`, because to me that would indicate an absolute, not relative path. The current code in this branch seems to handle that non-root situation just fine:

```python
In [21]: dt1 = xr.DataTree.from_dict({"root/c/a": xr.Dataset({"x": 1}), "root/c/b": xr.Dataset({"x": 2})})

In [22]: dt2 = xr.DataTree.from_dict(
    ...:     {"root/c/a": xr.Dataset({"x": 10}), "root/c/b": xr.Dataset({"x": 20})}
    ...: )

In [23]: result = {}

In [24]: for path, (node1, node2) in xr.group_subtrees(dt1['root'], dt2['root']):
    ...:     print(path)
    ...:     print(node1.dataset)
    ...:     result[path] = node1.dataset + node2.dataset
    ...: 
.
<xarray.DatasetView> Size: 0B
Dimensions:  ()
Data variables:
    *empty*
c
<xarray.DatasetView> Size: 0B
Dimensions:  ()
Data variables:
    *empty*
c/a
<xarray.DatasetView> Size: 8B
Dimensions:  ()
Data variables:
    x        int64 8B 1
c/b
<xarray.DatasetView> Size: 8B
Dimensions:  ()
Data variables:
    x        int64 8B 2
```

That looks good to me - the returned paths are clearly relative to `dt1['root']`, not the actual root (`dt`). 

> I guess "." might make more sense since it's the canonical path returned by NodePath?

I think it should be `.`, because its generally nice and intuitive to match the behaviour of `pathlib` (xref https://github.com/pydata/xarray/issues/9448):

```python
In [1]: from pathlib import PurePath

In [2]: PurePath('a/b/').relative_to('a/b/')
Out[2]: PurePosixPath('.')

In [3]: PurePath('a/b/').relative_to('a/')
Out[3]: PurePosixPath('b')

In [4]: PurePath('a/').relative_to('a/b', walk_up=True)  # requires python 3.12
Out[4]: PurePosixPath('..')

In [5]: PurePath('a/b/').relative_to('.')
Out[5]: PurePosixPath('a/b')
```
`NodePath` just inherits that behaviour.

TomNicholas replied:
I can also see an argument for `'./'`, because then direct string concatenation with `a/b` gives a valid relative path.

TomNicholas commented on xarray/tests/test_datatree_mapping.py:
```diff
@@ -184,100 +69,80 @@ def minmax(ds):
 
-        @map_over_datasets
-        def bad_func(ds):
-            # Datasets in simple_datatree have different numbers of dims
-            # TODO need to instead return different numbers of Dataset objects for this test to catch the intended error
-            return tuple(ds.dims)
+        # Datasets in simple_datatree have different numbers of dims
+        # TODO need to instead return different numbers of Dataset objects fo
+        # this test to catch the intended error
 
         with pytest.raises(TypeError, match="instead returns"):
-            bad_func(dt1)
+            map_over_datasets(lambda ds: tuple(ds.dims), dt1)
```
This change should make this test work now shouldn't it? So it can be un-xfailed?

shoyer replied:
done

TomNicholas reviewed (approved):
Great! I have a large number of small comments. The most important one is that there is an [unexpected error](https://xray--9636.org.readthedocs.build/en/9636/user-guide/hierarchical-data.html#iterating-over-multiple-trees) in one of the docs examples.

We should discuss if there is much more to do before releasing now.

TomNicholas commented:
I'm happy with it now that we're returning `"."`!

shoyer commented:
Submitting this so it doesn't go stale! Let's continue iterating on these ideas :)
